File: ziraffe.js
'use strict'

const createArray = (name1, name2) => {
    const array = new Array(name1.length + 1)
    for(let i = 0; i <= name1.length; i++){
        array[i] = new Array(name2.length + 1).fill(0)
        array[i][0] = i
    }
    for(let j = 0; j <= name2.length; j++) array[0][j] = j
    return array
}

const findDefaultFunc = (orig, defaultFunc) => {
    return orig === undefined? defaultFunc: orig
}

const restRequirements = (course, credits) => {
    return course.requirements.filter(item => credits.indexOf(item) < 0)
}
const gotRequirements = (course, credits) => {
    return course.requirements.filter(item => credits.indexOf(item) > 0)
}

module.exports = class Ziraffe {
    constructor() {
        this.lectures = require('./data/lectures').lectures
        this.courses = require('./data/courses').courses
    }

    gotCredits(course, credits) {
        let sum = 0
        credits.forEach(name => {
            if (course.requirements.indexOf(name) >= 0) {
                this.findLecture(name, (lecture) => {
                    sum = sum + lecture.credit
                })
            }
        })
        return sum
    }

    checkEachCourse(course, credits) {
        return {
            name: course.name,
            requirements: course.requirements,
            got_credits: this.gotCredits(course, credits),
            got_requirements: gotRequirements(course, credits),
            rest_requirements: restRequirements(course, credits),
        }
    }

    graduationCheck(credits) {
        const available = credits.filter(name => this.isFindLecture(name))
        return this.courses.map(course => this.checkEachCourse(course, available))
    }

    findLecture(name, success, failure) {
        const array = this.lectures.reduce((accumulator, current) => {
            if (current.name === name) {
                accumulator.push(current)
            }
            return accumulator
        }, [])
        if (array.length == 1) {
            success(array[0])
        } else {
            failure(name)
        }
    }

    isFindLecture(name) {
        const array = this.lectures.reduce((accumulator, current) => {
            if (current.name === name) {
                accumulator.push(current)
            }
            return accumulator
        }, [])
        return array.length == 1
    }

    similarLectures(name) {
        const array = this.lectures.map(item => this.calculateEditDistance(item.name, name))
        const array1 = array.filter(item => item.distance <= 1)
        if (array1.length == 0) {
            return array.filter(item => item.distance <= 2)
        }
        return array1
    }

    calculateEditDistance(name1, name2) {
        const distance = this.levenshtein(name1, name2)
        return { name1: name1, name2: name2, distance: distance }
    }

    levenshtein(name1, name2, insertFunc, removeFunc, updateFunc) {
        const array = createArray(name1, name2)
        insertFunc = findDefaultFunc(insertFunc, (node) => 1)
        removeFunc = findDefaultFunc(removeFunc, (node) => 1)
        updateFunc = findDefaultFunc(updateFunc, (string1, string2) => string1 !== string2? 1: 0)
        for(let i = 1; i <= name1.length; i++) {
            for(let j = 1; j <= name2.length; j++) {
                const d1 = array[i - 1][j] + removeFunc(name1.charAt(i - 1))
                const d2 = array[i][j - 1] + insertFunc(name2.charAt(j - 1))
                const d3 = updateFunc(name1.charAt(i - 1), name2.charAt(j - 1))
                array[i][j] = Math.min(d1, d2, d3)
            }
        }
        for(let i = 0; i <= name1.length; i++) {
            let line = ""
            for(let j = 0; j <= name2.length; j++) {
                let item = "    " + array[i][j]
                line = line + item.slice(-3)
            }
            console.log(line)
        }
        return array[name1.length][name2.length]
    }

}

File: ziraffe.go
package ziraffe

import (
	"sort"
	"strings"
)

/*
DataStore shows master of lectures and course data.
*/
type DataStore interface {
	Lectures() []Lecture
	Courses() []Course
	Init() error
}

/*
Grade means target grades of lectures.
*/
type Grade int

/*
Credit means credits of a lectures.
*/
type CreditCount int

/*
Lecture shows a lecture.
*/
type Lecture struct {
	Name   string      `json:"name"`
	Grade  Grade       `json:"grade"`
	Credit CreditCount `json:"credit"`
}

/*
Course shows requirements and recommended lectures of a course.
*/
type Course struct {
	Name          string      `json:"name"`
	DiplomaCredit CreditCount `json:"diploma-credit"`
	Requirements  []string    `json:"requirements"`
	Recommends    []string    `json:"recommends"`
}

type distance struct {
	distance int
	lecture  Lecture
}

type CourseDiplomaResult struct {
	Name             string
	Requirements     []string
	DiplomaCredit    CreditCount
	GotCredit        CreditCount
	GotRequirements  []string
	RestRequirements []string
}

type Ziraffe struct {
	Store DataStore
}

func NewZiraffe(ds DataStore) *Ziraffe {
	z := Ziraffe{Store: ds}
	return &z
}

func contains(slice []string, item string) bool {
	for _, element := range slice {
		if element == item {
			return true
		}
	}
	return false
}

func (z *Ziraffe) countNumberOfCredits(gotCredits []string, course Course) CreditCount {
	var sum CreditCount = 0
	for _, credit := range gotCredits {
		if contains(course.Requirements, credit) || contains(course.Recommends, credit) {
			lecture := z.FindLecture(credit)
			if lecture != nil {
				sum += lecture.Credit
			}
		}
	}
	return sum
}

func findRequirements(gotCredits []string, requirements []string, includeFunc func(flag bool) bool) []string {
	results := []string{}
	for _, r := range requirements {
		if includeFunc(contains(gotCredits, r)) {
			results = append(results, r)
		}
	}
	return results
}

func (z *Ziraffe) CheckCourse(gotCredits []string, course Course) CourseDiplomaResult {
	return CourseDiplomaResult{
		Name:             course.Name,
		Requirements:     course.Requirements,
		GotCredit:        z.countNumberOfCredits(gotCredits, course),
		DiplomaCredit:    course.DiplomaCredit,
		GotRequirements:  findRequirements(gotCredits, course.Requirements, func(flag bool) bool { return flag }),
		RestRequirements: findRequirements(gotCredits, course.Requirements, func(flag bool) bool { return !flag }),
	}
}

/*
FindCourses finds courses from name with partial matching.
*/
func (z *Ziraffe) FindCourses(name string) []Course {
	results := []Course{}
	if name == "" {
		return z.Store.Courses()
	}
	for _, c := range z.Store.Courses() {
		if strings.Contains(c.Name, name) {
			results = append(results, c)
		}
	}
	return results
}

func (z *Ziraffe) FindLecture(name string) *Lecture {
	for _, lecture := range z.Store.Lectures() {
		if lecture.Name == name {
			return &lecture
		}
	}
	return nil
}

func (z *Ziraffe) FindSimilarLectures(name string) []Lecture {
	distances := []distance{}
	for _, lecture := range z.Store.Lectures() {
		distances = append(distances, distance{distance: LevenshteinS(name, lecture.Name), lecture: lecture})
	}
	sort.Slice(distances, func(i, j int) bool {
		return distances[i].distance < distances[j].distance
	})
	min := distances[0].distance
	results := []Lecture{}
	if min == 0 {
		return results
	}
	for _, d := range distances {
		if d.distance == min {
			results = append(results, d.lecture)
		} else {
			break
		}
	}
	return results
}
